tss-esapi/src/ffi.rs

```diff
@@ -0,0 +1,27 @@
+// Copyright 2022 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+pub(crate) mod data_zeroize;
+
+use crate::ffi::data_zeroize::FfiDataZeroize;
+use mbox::MBox;
+use std::ops::Deref;
+
+/// Function that takes ownership of data that has been
+/// allocated with C memory allocation functions in TSS while also
+/// zeroizing the memory before freeing it.
+///
+/// # Arguments
+/// * `ffi_data_ptr` - A pointer to the FFI data.
+///
+/// # Returns
+/// The owned version of the FFI data.
+pub(crate) fn to_owned_with_zeroized_source<T>(ffi_data_ptr: *mut T) -> T
+where
+    T: FfiDataZeroize + Copy,
+{
+    let mut ffi_data = unsafe { MBox::from_raw(ffi_data_ptr) };
+    let owned_ffi_data: T = *ffi_data.deref();
+    ffi_data.ffi_data_zeroize();
+    owned_ffi_data
+}
```

Comment:
I keep forgetting - why do we need to copy here instead of just returning the data that was behind the pointer? Is it not possible to take proper ownership of it?

Reply:
> I keep forgetting - why do we need to copy here instead of just returning the data that was behind the pointer? Is it not possible to take proper ownership of it?

The MBox has the into_inner function which would allow us to immediately take ownership of the data and free the source. Unfortunately it does not allow us to zeroize the source. So instead here copy the data to the stack then zeroize the mermory pointed to by the MBox(i.e. the source).

Reply:
I think this is where I always get confused, because I expect `MBox` to take ownership by just wrapping the source pointer into an owned object, instead of copying it and freeing the source.

Reply:
Yes, this might not be the best way of doing it. Maybe we should implement ```TryFrom<MBox<T>>``` for some of our types.

Reply:
We can use TryFrom with MBox but it does not solve the problem of zeroizing the source allocated by tpm2-tss before freeing it. So I think this is currently the way forward. Will add the handling of other sensitive data in other PRs.

Reply:
```TryFrom``` for ```MBox``` did not really improve things. So I am happy with this for now.